apps/prairielearn/src/ee/pages/instructorAiGenerateDraftEditor/RichTextEditor/index.tsx

```diff
@@ -0,0 +1,153 @@
+/* eslint perfectionist/sort-sets: error */
+
+import { Blockquote } from '@tiptap/extension-blockquote';
+import { Bold } from '@tiptap/extension-bold';
+import { BulletList } from '@tiptap/extension-bullet-list';
+import { Code } from '@tiptap/extension-code';
+import { Document } from '@tiptap/extension-document';
+import { Dropcursor } from '@tiptap/extension-dropcursor';
+import { Gapcursor } from '@tiptap/extension-gapcursor';
+import { Heading } from '@tiptap/extension-heading';
+import { Italic } from '@tiptap/extension-italic';
+import { ListItem, ListKeymap, OrderedList } from '@tiptap/extension-list';
+import { Paragraph } from '@tiptap/extension-paragraph';
+import { Strike } from '@tiptap/extension-strike';
+import { Subscript } from '@tiptap/extension-subscript';
+import { Superscript } from '@tiptap/extension-superscript';
+import { Text } from '@tiptap/extension-text';
+import { Underline } from '@tiptap/extension-underline';
+import { Focus, Selection, UndoRedo } from '@tiptap/extensions';
+import { EditorContent, useEditor } from '@tiptap/react';
+// import { BubbleMenu, FloatingMenu } from '@tiptap/react/menus';
+import { useState } from 'preact/compat';
+import prettierHtmlPlugin from 'prettier/plugins/html';
+import prettier from 'prettier/standalone';
+import { Card, Form } from 'react-bootstrap';
+
+import { PlQuestionPanel } from './extensions/pl-question-panel.js';
+import { RawHtml } from './extensions/raw-html.js';
+
+function formatHtmlWithPrettier(html: string): Promise<string> {
+  return prettier.format(html, {
+    parser: 'html',
+    plugins: [prettierHtmlPlugin],
+    printWidth: 100,
+    tabWidth: 2,
+  });
+}
+
+/**
+ * The main rich text editor component.
+ * @param params
+ * @param params.htmlContents - The initial HTML contents of the editor.
+ * @param params.csrfToken
+ */
+const RichTextEditor = ({
+  csrfToken: _csrfToken,
+  htmlContents,
+}: {
+  htmlContents: string | null;
+  csrfToken: string;
+}) => {
+  const editor = useEditor({
+    editorProps: {
+      attributes: {
+        'aria-label': 'Main content area, start typing to enter text.',
+        autocapitalize: 'off',
+        autocomplete: 'off',
+        autocorrect: 'off',
+      },
+    },
+    emitContentError: true,
+    enableContentCheck: true,
+    immediatelyRender: false,
+    onContentError: (event) => {
+      throw new Error(event.error.message);
+    },
+    onCreate: async ({ editor }) => {
+      const formattedHtml = await formatHtmlWithPrettier(editor.getHTML());
+      setFormattedHtml(formattedHtml);
+    },
+    onUpdate: async ({ editor }) => {
+      const formattedHtml = await formatHtmlWithPrettier(editor.getHTML());
```

Comment:
😬 I fear that this means we'll need an HTML formatter that's aware of our own special flavor of HTML, with e.g. `<markdown>`, `<pl-code>` being indentation-sensitive, Mustache, and so on. This is a problem to solve later, I guess. Maybe Prettier is the wrong tool for the job here? 🤷

Reply:
I'll call this out.